source/common/router/rds_impl.cc

```diff
@@ -115,35 +107,79 @@ void RdsRouteConfigProviderImpl::onConfigUpdate(const ResourceVector& resources,
     throw EnvoyException(fmt::format("Unexpected RDS configuration (expecting {}): {}",
                                      route_config_name_, route_config.name()));
   }
+
   const uint64_t new_hash = MessageUtil::hash(route_config);
   if (!config_info_ || new_hash != config_info_.value().last_config_hash_) {
-    ConfigConstSharedPtr new_config(new ConfigImpl(route_config, factory_context_, false));
     config_info_ = {new_hash, version_info};
+    route_config_proto_ = route_config;
     stats_.config_reload_.inc();
     ENVOY_LOG(debug, "rds: loading new configuration: config_name={} hash={}", route_config_name_,
               new_hash);
-    tls_->runOnAllThreads(
-        [this, new_config]() -> void { tls_->getTyped<ThreadLocalConfig>().config_ = new_config; });
-    route_config_proto_ = route_config;
+    for (auto* provider : route_config_providers_) {
+      provider->onConfigUpdate();
+    }
   }
+
   runInitializeCallbackIfAny();
 }
 
-void RdsRouteConfigProviderImpl::onConfigUpdateFailed(const EnvoyException*) {
+void RdsRouteConfigSubscription::onConfigUpdateFailed(const EnvoyException*) {
   // We need to allow server startup to continue, even if we have a bad
   // config.
   runInitializeCallbackIfAny();
 }
 
-void RdsRouteConfigProviderImpl::runInitializeCallbackIfAny() {
+void RdsRouteConfigSubscription::registerInitTarget(Init::Manager& init_manager) {
+  init_manager.registerTarget(*this);
+}
+
+void RdsRouteConfigSubscription::runInitializeCallbackIfAny() {
   if (initialize_callback_) {
     initialize_callback_();
     initialize_callback_ = nullptr;
   }
 }
 
-void RdsRouteConfigProviderImpl::registerInitTarget(Init::Manager& init_manager) {
-  init_manager.registerTarget(*this);
+RdsRouteConfigProviderImpl::RdsRouteConfigProviderImpl(
+    RdsRouteConfigSubscriptionSharedPtr&& subscription,
+    Server::Configuration::FactoryContext& factory_context)
+    : subscription_(std::move(subscription)), factory_context_(factory_context),
+      tls_(factory_context.threadLocal().allocateSlot()) {
+  ConfigConstSharedPtr initial_config;
+  if (subscription_->config_info_.has_value()) {
+    initial_config =
+        std::make_shared<ConfigImpl>(subscription_->route_config_proto_, factory_context_, false);
+  } else {
+    initial_config = std::make_shared<NullConfigImpl>();
+  }
+  tls_->set([initial_config](Event::Dispatcher&) -> ThreadLocal::ThreadLocalObjectSharedPtr {
+    return std::make_shared<ThreadLocalConfig>(initial_config);
+  });
+  subscription_->route_config_providers_.insert(this);
```

Comment:
Subscription defines a OnUpdate interface,  it only stores the interface pointers.

Reply:
No it won't work, you will need a pointer to RdsRouteConfigProviderImpl to implement `std::vector<RouteConfigProviderSharedPtr> getRdsRouteConfigProviders()` in `RouteConfigProviderManagerImpl`

All the onUpdate is only used in `rds_impl` and should not be used elsewhere, would rather keep this in private.